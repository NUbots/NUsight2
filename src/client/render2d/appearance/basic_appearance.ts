import { observable } from 'mobx'

type Fill = { color: string, alpha: number }
type Stroke = { color: string, alpha: number, width: number }

export type BasicAppearanceOpts = {
  fill?: Partial<Fill>,
  stroke?: Partial<Stroke>
}

export class BasicAppearance {
  @observable fill?: Fill
  @observable stroke?: Stroke

  constructor({ fill, stroke }: { fill?: Fill, stroke?: Stroke }) {
    this.fill = fill
    this.stroke = stroke
  }

  static of({ fill, stroke }: BasicAppearanceOpts = {}): BasicAppearance {
    return new BasicAppearance({
      fill: fill && { color: '#000000', alpha: 1, ...fill },
      stroke: stroke && { color: '#000000', alpha: 1, width: 1, ...stroke },
    })
  }
}
