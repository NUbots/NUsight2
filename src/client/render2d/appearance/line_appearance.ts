import { observable } from 'mobx'


export type LineAppearanceOpts = {
  stroke?: {
    color: string,
    alpha?: number,
    width?: number,
    cap?: 'butt' | 'round' | 'square'
    dashOffset?: number
    join?: 'bevel' | 'round' | 'miter'
  }
}

export class LineAppearance {
  @observable stroke: {
    color: string,
    alpha: number,
    width: number,
    cap: 'butt' | 'round' | 'square'
    dashOffset: number
    join: 'bevel' | 'round' | 'miter'
  } = {
    color: '#000000',
    alpha: 1,
    width: 1,
    cap: 'butt',
    dashOffset: 0,
    join: 'miter',
  }

  constructor(opts: LineAppearanceOpts) {
    Object.assign(this.stroke, opts.stroke)
  }

  static of(opts: LineAppearanceOpts = {}): LineAppearance {
    return new LineAppearance(opts)
  }
}
