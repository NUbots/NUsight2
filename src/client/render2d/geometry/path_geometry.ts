import { observable } from 'mobx'

import { Vector2 } from '../../math/vector2'

export type Path = Vector2

export class PathGeometry {
  @observable path: Path[]

  constructor(opts: PathGeometry) {
    this.path = opts.path
  }

  static of(path: Path[]): PathGeometry {
    return new PathGeometry({ path })
  }
}
