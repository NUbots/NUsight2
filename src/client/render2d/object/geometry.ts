import { ArcGeometry } from '../geometry/arc_geometry'
import { ArrowGeometry } from '../geometry/arrow_geometry'
import { CircleGeometry } from '../geometry/circle_geometry'
import { LineGeometry } from '../geometry/line_geometry'
import { PathGeometry } from '../geometry/path_geometry'
import { PolygonGeometry } from '../geometry/polygon_geometry'
import { TextGeometry } from '../geometry/text_geometry'

export type Geometry =
  ArcGeometry
  | ArrowGeometry
  | CircleGeometry
  | LineGeometry
  | PathGeometry
  | PolygonGeometry
  | TextGeometry
