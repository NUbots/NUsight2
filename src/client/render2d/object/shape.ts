import { observable } from 'mobx'

import { Appearance } from '../appearance/appearance'
import { BasicAppearance } from '../appearance/basic_appearance'

import { Geometry } from './geometry'

export type ShapeOpts<T extends Geometry> = {
  appearance: Appearance
  geometry: T
}

export class Shape<T extends Geometry> {
  @observable appearance: Appearance
  @observable geometry: T

  constructor(opts: ShapeOpts<T>) {
    this.appearance = opts.appearance
    this.geometry = opts.geometry
  }

  static of<T extends Geometry>(geometry: T, appearance: Appearance = BasicAppearance.of()) {
    return new Shape<T>({
      appearance,
      geometry,
    })
  }
}
