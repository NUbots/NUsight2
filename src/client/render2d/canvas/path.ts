import { PathGeometry } from '../geometry/path_geometry'
import { Shape } from '../object/shape'

import { applyAppearance } from './rendering'

export function renderPath(ctx: CanvasRenderingContext2D, shape: Shape<PathGeometry>): void {
  const { points } = shape.geometry
  const start = points[0]

  ctx.beginPath()
  ctx.moveTo(start.x, start.y)
  points.forEach(p => ctx.lineTo(p.x, p.y))

  applyAppearance(ctx, shape.appearance)

  ctx.stroke()
}
