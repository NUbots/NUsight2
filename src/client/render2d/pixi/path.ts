import { createTransformer } from 'mobx-utils'
import { Graphics } from 'pixi.js'

import { PathGeometry } from '../geometry/path_geometry'
import { Shape } from '../object/shape'

import { applyAppearance } from './rendering'

export const renderPath = createTransformer((shape: Shape<PathGeometry>): Graphics => {

  const { points } = shape.geometry
  const start = points[0]

  const g = new Graphics()

  applyAppearance(g, shape.appearance, g => {
    g.moveTo(start.x, start.y)
    points.forEach(p => g.lineTo(p.x, p.y))
  })

  return g
})
