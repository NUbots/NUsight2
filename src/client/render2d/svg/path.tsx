import { observer } from 'mobx-react'
import * as React from 'react'

import { Transform } from '../../math/transform'
import { PathGeometry } from '../geometry/path_geometry'
import { Shape } from '../object/shape'

import { toSvgProps } from './rendering'

type Props = { model: Shape<PathGeometry>, world: Transform }
export const Path = observer(({ model: { geometry: { points }, appearance } }: Props) => {
  const start = points[0]
  const path = `M${start.x},${start.y}` + points.map(p => `L${p.x},${p.y}`)

  return (
    <path
      d={path}
      {...toSvgProps(appearance)}
      fill='none'
    />
  )
})
