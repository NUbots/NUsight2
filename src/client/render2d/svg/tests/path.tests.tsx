import * as React from 'react'

import { Transform } from '../../../math/transform'
import { BasicAppearance } from '../../appearance/basic_appearance'
import { PathGeometry } from '../../geometry/path_geometry'
import { Shape } from '../../object/shape'
import { Path } from '../path'

describe('PathSVGRenderer', () => {
  it('renders', () => {
    const model = Shape.of(PathGeometry.of([]), BasicAppearance.of())
    const world = Transform.of()
    expect(<Path model={model} world={world}/>).toMatchSnapshot()
  })
})
