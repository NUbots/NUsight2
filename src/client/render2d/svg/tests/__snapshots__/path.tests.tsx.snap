// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PathSVGRenderer renders 1`] = `
<_class
  model={
    Shape {
      "appearance": BasicAppearance {
        "fill": undefined,
        "stroke": undefined,
      },
      "geometry": PathGeometry {
        "points": Array [],
      },
    }
  }
  world={
    Transform {
      "anticlockwise": true,
      "rotate": 0,
      "scale": Object {
        "x": 1,
        "y": 1,
      },
      "translate": Object {
        "x": 0,
        "y": 0,
      },
    }
  }
/>
`;
